storey/sources.py

```diff
@@ -730,212 +754,80 @@ def __init__(
         if parse_dates:
             if not isinstance(parse_dates, List):
                 parse_dates = [parse_dates]
-            if isinstance(parse_dates, List):
-                if self._with_header and any([isinstance(f, int) for f in self._parse_dates]):
-                    raise ValueError("parse_dates can be list of int only when there is no header")
-                if not self._with_header and all([isinstance(f, int) for f in self._parse_dates]):
-                    self._dates_indices = parse_dates
-        if isinstance(self._time_field, int):
-            if self._with_header:
-                raise ValueError("time field can be int only when there is no header")
+            if not self._with_header and any([isinstance(f, str) for f in self._parse_dates]):
+                raise ValueError("parse_dates can be list of str only when there is header")
+            self._dates_indices.extend(parse_dates)
+        if self._time_field is not None:
+            if not header and isinstance(self._time_field, str):
+                raise ValueError("time_field can only be set to an integer when with_header is false")
             self._dates_indices.append(self._time_field)
-
-        if not header and isinstance(key_field, str):
+        if not header and (
+            isinstance(key_field, str)
+            or (isinstance(key_field, list) and any([isinstance(key, str) for key in key_field]))
```

Comment:
I really hope we're not accepting lists of mixed types here... It's probably enough to test just the first element.

Reply:
I disagree, I think we should check every key.

Reply:
we can get mixed types with header =True as before. We need to talk about that...

Reply:
Well, the key field list comes from the user. We don't have to support a mixed list of strings and ints. That's not useful to the user and only adds needless complexity to our code.

Reply:
added a mixed type validator in _validate_fields.I do not think I should add a test for first element because in that solution I will need also to check if the list is not empty.